File: 914.X-of-a-Kind-in-a-Deck-of-Cards.go
// https://leetcode.com/problems/x-of-a-kind-in-a-deck-of-cards/description/
//
// algorithms
// Easy (33.08%)
// Total Accepted:    4.3k
// Total Submissions: 13k
// beats 100.0% of golang submissions
// 这道题首先需要得到slice中每个元素的个数，然后从2开始，直到slice的最小值
// 如果map的value都能够整除base，那么说明能够按照X=base进行split，否则，当
// base > min的时候，就不需要继续寻找了，肯定没有合适的X

package leetcode

func hasGroupsSizeX(deck []int) bool {
	cardMap := make(map[int]int)
	length := len(deck)
	if length == 1 {
		return false
	}

	for _, n := range deck {
		cntNum, ok := cardMap[n]
		if ok {
			cardMap[n] = cntNum + 1
		} else {
			cardMap[n] = 1
		}
	}

	minCard := int(^uint(0) >> 1)
	for _, v := range cardMap {
		if v < minCard {
			minCard = v
		}
	}

	for base := 2; base <= minCard; base++ {
		isFound := true
		for _, v := range cardMap {
			if v%base != 0 {
				isFound = false
				break
			}
		}
		if isFound {
			return true
		}
	}

	return false
}

File: 914.X-of-a-Kind-in-a-Deck-of-Cards.java
// https://leetcode.com/problems/x-of-a-kind-in-a-deck-of-cards/
//
// algorithms
// Easy (33.4%)
// Total Accepted: 24,961
// Total Submissions: 73,423


class Solution {
    public boolean hasGroupsSizeX(int[] deck) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int n : deck) {
            map.put(n, map.getOrDefault(n, 0) + 1);
        }

        int base = -1, gcd = -1;
        for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
            int value = entry.getValue();
            if (value < 2) {
                return false;
            }

            if (base == -1) {
                base = value;
            } else if ((gcd = getGcd(base, value)) > 1) {
                base = gcd;
            } else {
                return false;
            }
        }

        return true;
    }

    private int getGcd(int x, int y) {
        int max, min;
        if (x > y) {
            max = x;
            min = y;
        } else {
            max = y;
            min = x;
        }

        if (max % min != 0) {
            return getGcd(min, max % min);
        }

        return min;
    }

}